Release assertion failure in LegacyRenderSVGShape::strokeBoundingBox()

rwlbuis opened the pull request:
#### 6866936923f71c78e121c3f98c950ada4a991841
<pre>
Release assertion failure in LegacyRenderSVGShape::strokeBoundingBox()
<a href="https://bugs.webkit.org/show_bug.cgi?id=285429">https://bugs.webkit.org/show_bug.cgi?id=285429</a>

Reviewed by Said Abou-Hallawa.

The test case converts the SVG baseVal x/y values on the filter from a percentage to a
user unit. To do this convertValueFromUserUnitsToPercentage is used, but the divisor
in the calculation will be zero since the filter width/height is zero (due to empty outer svg viewport), so in the
end SVG baseVal x/y values will be set to NaN.

This causes a problem on when calculating the filter repaint rect in SVGRenderSupport::intersectRepaintRectWithResources
since the x and y values are NaN, causing m_accurateRepaintBoundingBox to be an empty Markable object and thus the
Release assertion failure upon dereference.

To fix this detect whenever the divisor is zero in convertValueFromUserUnitsToPercentage and return 0 if so.

* LayoutTests/svg/crash-svg-filter-empty-viewport-expected.txt: Added.
* LayoutTests/svg/crash-svg-filter-empty-viewport.svg: Added.
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::convertValueFromUserUnitsToPercentage const):

Canonical link: <a href="https://commits.webkit.org/289664@main">https://commits.webkit.org/289664@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b0386a4f1498690a13458993d31ec9a5da4c4c0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87544 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7057 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41919 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92406 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38288 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89595 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7439 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15226 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67616 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25366 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90546 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5683 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79222 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47968 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5469 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33616 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37399 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75876 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34480 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94291 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14709 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10793 "Found 1 new test failure: imported/w3c/web-platform-tests/html/capability-delegation/delegate-fullscreen-request-subframe-same-origin.https.tentative.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76451 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14963 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75071 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75671 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18632 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20051 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18476 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7663 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14727 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20028 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14472 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17914 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16253 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on LayoutTests/svg/crash-svg-filter-empty-viewport.svg:
```diff
@@ -0,0 +1,18 @@
+<svg width="0" height="0" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" onload="test()">
```
No need for `xmlns:xlink`. This namespace was deprecated in SVG2.

shallawa replied:
Does the SVG have to have zero size `width="0" height="0"` to reproduce the bug?

rwlbuis replied:
> No need for `xmlns:xlink`. This namespace was deprecated in SVG2.

Removed.

rwlbuis replied:
> Does the SVG have to have zero size `width="0" height="0"` to reproduce the bug?

Yes, this causes an empty viewport and thus the filter x/y percentages are calculated against a 0x0 viewport and the divisor becomes zero.

shallawa commented on LayoutTests/svg/crash-svg-filter-empty-viewport.svg:
```diff
@@ -0,0 +1,18 @@
+    function test() {
+      let filter = document.getElementById("filter");
+      let length = filter.x.baseVal;
+      length.value = 1;
```
Should not we verify the value of length after setting it?
```
shouldBe("length.unitType", "SVGAngle.SVG_LENGTHTYPE_PERCENTAGE");
shouldBe("length.value", "1");
```

rwlbuis replied:
I added that, but because the value is calculated against the empty viewport length.value ends up being zero, matching what Chrome does.

shallawa commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -198,7 +198,10 @@ ExceptionOr<float> SVGLengthContext::convertValueFromUserUnitsToPercentage(float
+    if (auto divisor = dimensionForLengthMode(lengthMode, *viewportSize))
+        return value / divisor * 100;
+
+    return 0;
```
Should not we return `value` instead of zero?

The specs of https://svgwg.org/svg2-draft/types.html#InterfaceSVGLength says:

Otherwise, if unitType is [SVG_LENGTHTYPE_PERCENTAGE](https://svgwg.org/svg2-draft/types.html#__svg__SVGLength__SVG_LENGTHTYPE_PERCENTAGE), then:
Let viewport size be a basis to resolve percentages against, based on the [SVGLength](https://svgwg.org/svg2-draft/types.html#InterfaceSVGLength)'s [associated element](https://svgwg.org/svg2-draft/types.html#LengthAssociatedElement) and [directionality](https://svgwg.org/svg2-draft/types.html#LengthAssociatedElement):
has no [associated element](https://svgwg.org/svg2-draft/types.html#LengthAssociatedElement)
size is 100

So `dimensionForLengthMode(lengthMode, *viewportSize)` should return 100 in this case. This means we should return `value / 100 * 100` or `value`.

rwlbuis replied:
I changed it to return value.

shallawa commented on LayoutTests/svg/crash-svg-filter-empty-viewport.svg:
```diff
@@ -0,0 +1,27 @@
+  <filter id="filter"></filter>
+  <marker id="marker" orient="1" filter="url(#filter)"></marker>
+  <line marker-end="url(#marker)"></line>
+  <text>PASS</text>
```
I think this `<text>` can be removed.

rwlbuis replied:
Removed.

shallawa commented on LayoutTests/svg/crash-svg-filter-empty-viewport.svg:
```diff
@@ -0,0 +1,27 @@
+    function test() {
+      let filter = document.getElementById("filter");
+      x = filter.x.baseVal;
+      shouldBe('x.unitType', "SVGLength.SVG_LENGTHTYPE_PERCENTAGE");
```
Please add a `description('...');` statement here to describe what this test is doing.

rwlbuis replied:
Done.

shallawa commented on LayoutTests/svg/crash-svg-filter-empty-viewport.svg:
```diff
@@ -0,0 +1,27 @@
+      let filter = document.getElementById("filter");
+      x = filter.x.baseVal;
+      shouldBe('x.unitType', "SVGLength.SVG_LENGTHTYPE_PERCENTAGE");
+      shouldBe('x.value', "-0");
```
Do you know why the value is `-0`? Is this because the initial value of the `x` property of the `<filter>` element is `-10%`?

rwlbuis replied:
Indeed it is. Note that Chrome has the same behaviour.

rwlbuis committed:
Release assertion failure in LegacyRenderSVGShape::strokeBoundingBox()
https://bugs.webkit.org/show_bug.cgi?id=285429

Reviewed by Said Abou-Hallawa.

The test case converts the SVG baseVal x/y values on the filter from a percentage to a
user unit. To do this convertValueFromUserUnitsToPercentage is used, but the divisor
in the calculation will be zero since the filter width/height is zero (due to empty outer svg viewport), so in the
end SVG baseVal x/y values will be set to NaN.

This causes a problem on when calculating the filter repaint rect in SVGRenderSupport::intersectRepaintRectWithResources
since the x and y values are NaN, causing m_accurateRepaintBoundingBox to be an empty Markable object and thus the
Release assertion failure upon dereference.

To fix this detect whenever the divisor is zero in convertValueFromUserUnitsToPercentage and return 0 if so.

* LayoutTests/svg/crash-svg-filter-empty-viewport-expected.txt: Added.
* LayoutTests/svg/crash-svg-filter-empty-viewport.svg: Added.
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::convertValueFromUserUnitsToPercentage const):

Canonical link: https://commits.webkit.org/289664@main
